# Pythonにも条件付きfor文が欲しい

これね
```cpp
for (int i = 2; i*i <= N; i++) {
    // 処理
}
```

In [33]:
N = 100000

In [34]:
%%timeit

cnt = 0
for i in range(N):
    cnt += i

3.12 ms ± 49.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [35]:
%%timeit

cnt = 0
i = 0
while i <= N:
    cnt += i
    i += 1

5.67 ms ± 143 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [36]:
def range_while(start, condition):
    while condition(start):
        yield start
        start += 1

In [37]:
%%timeit

cnt = 0
for i in range_while(0, lambda i: i <= N):
    cnt += i

10.4 ms ± 228 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## 処理速度比較

| loop-type | 10^5 | 10^7 |
| :-------: | :--: | :--: |
| for loop | 3.12 ms ± 49.6 µs | 31.2 ms ± 488 µs |
| while loop | 5.67 ms ± 143 µs | 55.3 ms ± 486 µs |
| range_while | 10.4 ms ± 228 µs | 108 ms ± 3.94 ms |

流石にちょっと遅いなあ